docs/source/format/CDataInterface/PyCapsuleInterface.rst

```diff
@@ -0,0 +1,402 @@
+
+      .. code-block:: c
+
+         #include <Python.h>
+         
+         // If the capsule is not an ArrowSchema, will return NULL.
+         ArrowSchema* GetArrowSchemaPyCapsule(PyObject* capsule) {
+           return PyCapsule_GetPointer(schema, "arrow_schema");
+         }
+
+   .. tab-item:: Cython
+
+      .. code-block:: cython
+
+         import cpython
+        
+         cdef object get_arrow_schema_py_capsule(PyObject* capsule):
+             return cpython.PyCapsule_GetPointer(capsule, 'arrow_schema')
+
+Backwards Compatibility with PyArrow
+------------------------------------
+
+When interacting with PyArrow, the PyCapsule interface will be preferred over
+the ``_export_to_c`` and ``_import_from_c`` methods. However, many libraries will
+want to support a range of PyArrow versions. This can be done via Duck typing.
+
+For example, if your library had an import method such as:
+
+.. code-block:: python
+
+   # OLD METHOD
+   def from_arrow(arr: pa.Array)
+       array_import_ptr = make_array_import_ptr()
+       schema_import_ptr = make_schema_import_ptr()
+       arr._export_to_c(array_import_ptr, schema_import_ptr)
+       return import_c_data(array_import_ptr, schema_import_ptr)
+
+You can rewrite this method to support both PyArrow and other libraries that
+implement the PyCapsule interface:
+
+.. code-block:: python
+
+   # NEW METHOD
+   def from_arrow(arr)
+       if hasattr(arr, "__arrow_c_array__"):
+            schema_ptr, array_ptr = arr.__arrow_c_array__()
+            return import_c_capsule_data(schema_ptr, array_ptr)
+       elif isinstance(arr, pa.Array):
+            # Deprecated method, used for older versions of PyArrow
+            array_import_ptr = make_array_import_ptr()
+            schema_import_ptr = make_schema_import_ptr()
+            arr._export_to_c(array_import_ptr, schema_import_ptr)
+            return import_c_data(array_import_ptr, schema_import_ptr)
+       else:
+           raise TypeError(f"Cannot import {type(arr)} as Arrow array data.")
+
+You may also wish to accept objects implementing the protocol in your
+constructors. For example, in PyArrow, the :py:func:`array` and :py:func:`record_batch`
+constructor accepts any object that implements the ``ArrowArrayExportable``
+protocol. Similarly, the PyArrow's :py:func:`schema` constructor accepts any object
+that implements the ``ArrowSchemaExportable`` protocol.
+
+Now if your library has an export to PyArrow function, such as:
+
+.. code-block:: python
+
+   # OLD METHOD
+   def to_arrow(self) -> pa.Array:
+       array_export_ptr = make_array_export_ptr()
+       schema_export_ptr = make_schema_export_ptr()
+       self.export_c_data(array_export_ptr, schema_export_ptr)
+       return pa.Array._import_from_c(array_export_ptr, schema_export_ptr)
+
+You can implement the protocol instead and encourage users to prefer directly
+passing your object to the :py:func:`array` constructor, which accepts any object
+that implements the protocol:
```

Comment:
I am not sure we should recommend deprecating such explicit methods? (personally I am not sure I would do that if I were maintaining such a library, it's also nice (and more discoverable) to have an explicit method for it)

Reply:
Indeed, I would leave it to third-party libraries to make such an API decision.

Reply:
I've removed this.